### PR TITLE
Documented automatic plug-in installation and upgrade

### PR DESCRIPTION
**Pull request by stuclem**
Towards https://github.com/vmware/vic-product/issues/1941. Documented automatic install and upgrade of the plug-ins.

Also fixes https://github.com/vmware/vic-product/issues/1940.

### CODE REVIEW

**Comment by stuclem** [comment]
https://github.com/vmware/vic-product/pull/2012/commits/b2a9dd8e5642145a09979e6f8396380ed2be5742 fixes https://github.com/vmware/vic-product/issues/2037

**Comment by stuclem** [comment]
https://github.com/vmware/vic-product/pull/2012/commits/3288a995397a801b3123697b5abff5b1799a8458 is an attempt to fix https://github.com/vmware/vic-product/issues/1356, to handle deployments with multiple appliances on the same VC. However, I am unsure of the logic of the checkbox. https://github.com/vmware/vic-product/issues/1952#issuecomment-415301881 suggests that this checkbox was introduced to avoid automatically installing a version of the plug-in that conflicts with an existing appliance of another version. However, in https://github.com/vmware/vic-ui/issues/365#issuecomment-366288567 @jooskim suggests that one must always run the plug-in installer for every appliance that you deploy, even if the plug-in is already installed (although this comment predates the automatic installer and doesn't take different appliance versions into account).

@zjs @renmaosheng can you please clarify the circumstances under which one should deselect the **Install UI Plugin** checkbox?

**Comment by stuclem** [comment]
I think that this PR is now ready for review. @zjs and @renmaosheng can you please review, or assign an appropriate reviewer? Thanks!

**Comment by zjs** [comment]
I'll take a look, but I'm also adding @morris-jason, who implemented most of the feature, and @wjun, who finished it.

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md [hunk 1]
```diff
@@ -6,15 +6,18 @@ The following services run in the vSphere Integrated Containers appliance:
...
+- The file server for vSphere Integrated Containers Engine downloads and the installation of the vSphere Client plug-ins
+- The `vic-machine-server` service, that powers the vSphere Integrated Containers plug-in for the HTML5 vSphere Client.
 
-You can deploy multiple vSphere Integrated Containers appliances to the same vCenter Server instance. Also, if a Platform Services Controller manages multiple vCenter Server instances, you can deploy multiple appliances to different vCenter Server instances that share that Platform Services Controller.
+If you are deploying vSphere Integrated Containers 1.4.3 or later, by default the vSphere Integrated Containers plug-ins for the vSphere Client are installed automatically. The installer installs a basic plug-in for the Flex-based vSphere Web Client on vCenter Server 6.0, 6.5, or 6.7, and a plug-in with more complete functionality for the HTML5 vSphere Client on vCenter Server 6.5 and 6.7. If you are installing a version of vSphere Integrated Containers that pre-dates 1.4.3, you must install the plug-ins manually.
```
Comment (zjs) [inline docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: > The installer installs a basic plug-in for the Flex-based vSphere Web Client on vCenter Server 6.0, 6.5, or 6.7

I don't believe we install the Flex-based plugin on 6.7.

https://github.com/vmware/vic-product/blob/d6e5e9895fc6c2ce943fe116745648c975bc8808/installer/fileserver/tasks/plugin.go#L426-L447

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md [hunk 2]
```diff
@@ -100,7 +103,14 @@ You can deploy multiple vSphere Integrated Containers appliances to the same vCe
...
 
-12. Click **Continue** to initialize the appliance.
+12. To automatically install the vSphere Integrated Containers plug-in for vSphere Client, leave the **Install UI Plugin** check box selected.
+
+    **NOTES**: 
+    - The option to automatically install the  plug-ins for the vSphere Client is available in vSphere Integrated Containers 1.4.3 and later. 
+    - By default, in an environment in which multiple vSphere Integrated Containers are deployed to the same vCenter Server instance, the vSphere Integrated Containers plug-in connects to whichever appliance was most recently registered with vCenter Server. Consequently, if there are older instances of the appliance registered with vCenter Server and you do not want the plug-in to automatically upgrade to the latest version, deselect the **Install UI Plugin** check box.
```
Comment (zjs) [inline docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: > By default, in an environment in which multiple vSphere Integrated Containers are deployed to the same vCenter Server instance, the vSphere Integrated Containers plug-in connects to whichever appliance was most recently registered with vCenter Server.

Is this true? (I'm not sure what the current behavior is when multiple appliances are present.)

cc @wjun

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md [hunk 3]
```diff
@@ -100,7 +103,14 @@ You can deploy multiple vSphere Integrated Containers appliances to the same vCe
...
+    - If you deselect the **Install UI Plugin** check box, you can install or upgrade the plug-ins later. 
+    - If you are installing a version of vSphere Integrated Containers that pre-dates 1.4.3, you must install the plug-ins manually.
+13. Verify that the certificate thumbprint for vCenter Server is valid, and click **Continue** to initialize the appliance.
```
Comment (zjs) [inline docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: I think we have a page about how to verify thumprints. Should this link there?

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md (hunk 2 above)
Comment (wjun) [reply docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: The plugin behavior does not change in 1.4.3, so with multiple VIC instances existing, the plugin still pick up one instance with the previous algorithm which cannot guarantee the latest version to be selected. Thus, I think we could document like below:

in an environment in which multiple vSphere Integrated Containers are deployed to the same vCenter Server instance, the vSphere Integrated Containers plug-in connects to one appliance that may not be the most recently registered with vCenter Server. Consequently, if there are older instances of the appliance registered with vCenter Server and you do not want the plug-in to automatically upgrade to the latest version, deselect the **Install UI Plugin** check box.

File: docs/user_doc/vic_vsphere_admin/upgrade_appliance.md [hunk 4]
```diff
@@ -51,22 +53,25 @@ For information about the supported upgrade paths for all versions of vSphere In
...
 
     1. Enter the name of the datacenter that contains the old version of the appliance.
-    2. Enter the IP address of the old version of the appliance.  The upgrade script does not accept FQDN addresses for the old appliance.
+    2. Enter the IP address of the old version of the appliance.The upgrade script does not accept FQDN addresses for the old appliance.
     3. For the old appliance user name, enter `root`.
-    4. Enter the root password for the old appliance.
-6. Verify that the upgrade script has detected your upgrade path correctly.        
+8. To automatically upgrade the vSphere Integrated Containers plug-in for vSphere Client, enter `y` at the prompt to `Upgrade VIC UI Plugin`.
+
+    **NOTE**: The option to automatically upgrade the  plug-in for the vSphere Client is available in vSphere Integrated Containers 1.4.3 and later. If you you enter `n` to skip the plug-in upgrade, you can upgrade the plug-in later. If are upgrading to a version of vSphere Integrated Containers that pre-dates 1.4.3, you must upgrade the plug-in manually.
```
Comment (wjun) [inline docs/user_doc/vic_vsphere_admin/upgrade_appliance.md]: "In are upgrading to a version of vSphere Integrated Containers", missing 'you'

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md (hunk 1 above)
Comment (stuclem) [reply docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: @zjs and @wjun, having one plug-in for 6.0 and one for 6.5/6.7 makes things a lot simpler. Corrected the vSphere version and singularized the word "plug-ins" in https://github.com/vmware/vic-product/pull/2012/commits/0d64bf5fd00d6ce57027e219b9977731e2a84637.

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md (hunk 2 above)
Comment (stuclem) [reply docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: Fixed in https://github.com/vmware/vic-product/pull/2012/commits/1f2db4c89ff1260c6785ea08d6e299ca05126a9e.

File: docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md (hunk 3 above)
Comment (stuclem) [reply docs/user_doc/vic_vsphere_admin/deploy_vic_appliance.md]: We do so from the prereqs. Our editorial guidelines advise against including links in procedure steps, which cause people go off and look at other stuff and then forget what they were doing or where they were in the procedure.

File: docs/user_doc/vic_vsphere_admin/upgrade_appliance.md (hunk 4 above)
Comment (stuclem) [reply docs/user_doc/vic_vsphere_admin/upgrade_appliance.md]: Fixed in https://github.com/vmware/vic-product/pull/2012/commits/b2c586aacafe22bd8bff43054ea3d149bd62609e

**Comment by stuclem** [comment]
Thanks @wjun !

**Comment by stuclem** [comment]
@lgayatri needs the auto-install/upgrade docs for testing purposes. Merging. @morris-jason please feel free to make comments even after I have merged.

**Pull request merged**